lib/Runtime/Debug/TTActionEvents.cpp

```diff
@@ -0,0 +1,986 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+#include "RuntimeDebugPch.h"
+
+#if ENABLE_TTD
+
+namespace TTD
+{
+    namespace NSLogEvents
+    {
+        bool IsJsRTActionExecutedInScriptWrapper(EventKind tag)
+        {
+            switch(tag)
+            {
+            case EventKind::GetAndClearExceptionActionTag:
+                return false;
+            default:
+                return true;
+            }
+        }
+
+        bool IsJsRTActionRootCall(const EventLogEntry* evt)
+        {
+            if(evt->EventKind != NSLogEvents::EventKind::CallExistingFunctionActionTag)
+            {
+                return false;
+            }
+
+            const JsRTCallFunctionAction* cfAction = GetInlineEventDataAs<JsRTCallFunctionAction, EventKind::CallExistingFunctionActionTag>(evt);
+            return cfAction->CallbackDepth == 0;
+        }
+
+        int64 AccessTimeInRootCallOrSnapshot(const EventLogEntry* evt, bool& isSnap, bool& isRoot, bool& hasRtrSnap)
+        {
+            isSnap = false;
+            isRoot = false;
+            hasRtrSnap = false;
+
+            if(evt->EventKind == NSLogEvents::EventKind::SnapshotTag)
+            {
+                const NSLogEvents::SnapshotEventLogEntry* snapEvent = NSLogEvents::GetInlineEventDataAs<NSLogEvents::SnapshotEventLogEntry, NSLogEvents::EventKind::SnapshotTag>(evt);
+
+                isSnap = true;
+                return snapEvent->RestoreTimestamp;
+            }
+
+            if(NSLogEvents::IsJsRTActionRootCall(evt))
+            {
+                const NSLogEvents::JsRTCallFunctionAction* rootEntry = NSLogEvents::GetInlineEventDataAs<NSLogEvents::JsRTCallFunctionAction, NSLogEvents::EventKind::CallExistingFunctionActionTag>(evt);
+
+                isRoot = true;
+                hasRtrSnap = rootEntry->AdditionalInfo->RtRSnap != nullptr;
+                return rootEntry->AdditionalInfo->CallEventTime;
+            }
+
+            return -10;
```

Comment:
Switched to "natural" -1 value (any invalid value is ok as all callers check bool out params before use).